mysql/src/lib.rs

```diff
@@ -502,6 +502,14 @@ impl<B: AsyncMysqlShim<Cursor<Vec<u8>>> + Send + Sync, S: AsyncRead + AsyncWrite
                         let schema = ::std::str::from_utf8(&q[b"USE ".len()..])
                             .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
                         let schema = schema.trim().trim_end_matches(';').trim_matches('`');
+                        if schema.is_empty() {
+                            w.error(
+                                errorcodes::ErrorKind::ER_NO_DB_ERROR,
+                                "No database selected".as_bytes(),
+                            )?;
+                            self.writer_flush().await?;
+                            continue;
```

Comment:
Why need `continue`?

Reply:
> Why need `continue`?

in order to skip the following line:  `self.shim.on_init(schema, w).await?;`

Reply:
If schema is empty, could we return err in shim.on_init?

Like MySQL in the linked issue #4723

Reply:
I think we should let opensrv give the original `use` statement to databend-query, and the query to check the `use` is empty or not.

Reply:
could we add an `on_accept` method for `AsyncMysqlShim` trait in `opensrv/mysql/src/lib.rs` and call `on_accpet` when we receive an new connection so that we can know whether we are switching to an empty database.

code as below: 
```diff
// opensrv/mysql/src/lib.rs
if let Some(Ok(db)) = handshake.db.as_ref().map(|x| std::str::from_utf8(x)) {
        let w = InitWriter {
            client_capabilities: self.client_capabilities,
            writer: &mut self.writer,
        };
-       self.shim.on_init(db, w).await?;
+       self.shim.on_accept(db, w).await?;
```

Reply:
The backgroud is that `use <db>` will be used in two places:
1. When MySQL client connect to the databend-query, we can handle it in `on_init` https://github.com/datafuselabs/databend/blob/f25c34917b415ab42308a99637b5513aa4ea76df/query/src/servers/mysql/mysql_interactive_worker.rs#L196-L214
2. The user run the query `use <db>` manualy in the client, we can handle it in `do_query` https://github.com/datafuselabs/databend/blob/f25c34917b415ab42308a99637b5513aa4ea76df/query/src/servers/mysql/mysql_interactive_worker.rs#L265
So, let's the `use <db>` query bypass to the databend and check them in the above place looks a good idea.

Reply:
Actually, when database name is empty, it will return directy in `do_init` and `do_query` wont be called.

```
  async fn do_init(&mut self, database_name: &str) -> Result<()> {
        if database_name.is_empty() {
            return Ok(());
        }
        let init_query = format!("USE `{}`;", database_name);


        let do_query = self.do_query(&init_query).await;
        match do_query {
            Ok(_) => Ok(()),
            Err(error_code) => Err(error_code),
        }
    }
}
```

if we change it to 

```diff

async fn do_init(&mut self, database_name: &str) -> Result<()> {
-       if database_name.is_empty() {
-          return Ok(());
-       }
        let init_query = format!("USE `{}`;", database_name);


        let do_query = self.do_query(&init_query).await;
        match do_query {
            Ok(_) => Ok(()),
            Err(error_code) => Err(error_code),
        }
    }
}
```

and check database name in `do_query`, **new connection will fail because new connection can be made without database name**(can see the test case https://github.com/datafuselabs/databend/blob/main/tests/suites/0_stateless/10_drivers/10_0000_python_mysql_driver.py)

Reply:
Look this example:

```
>>> import mysql.connector
>>> mydb = mysql.connector.connect(
...   host="127.0.0.1",
...   user="root",
...   passwd="",
...   port="3306")
>>> 
>>> mycursor = mydb.cursor()
>>>  
>>> mycursor.execute("SHOW TABLES")
Traceback (most recent call last):
  File "/home/taichong/.local/lib/python3.8/site-packages/mysql/connector/connection_cext.py", line 523, in cmd_query
    self._cmysql.query(query,
_mysql_connector.MySQLInterfaceError: No database selected

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/taichong/.local/lib/python3.8/site-packages/mysql/connector/cursor_cext.py", line 269, in execute
    result = self._cnx.cmd_query(stmt, raw=self._raw,
  File "/home/taichong/.local/lib/python3.8/site-packages/mysql/connector/connection_cext.py", line 528, in cmd_query
    raise errors.get_mysql_exception(exc.errno, msg=exc.msg,
mysql.connector.errors.ProgrammingError: 1046 (3D000): No database selected
>>>  
```

If not set `db='xx'` the python driver will report a err when connect mysql.

So we could modify this test.

Reply:
Bye the way, I find other err:

In MySQL :

Reply:
> Actually, when database name is empty, it will return directy in `do_init` and `do_query` wont be called.
> 
> ```
>   async fn do_init(&mut self, database_name: &str) -> Result<()> {
>         if database_name.is_empty() {
>             return Ok(());
>         }
>         let init_query = format!("USE `{}`;", database_name);
> 
> 
>         let do_query = self.do_query(&init_query).await;
>         match do_query {
>             Ok(_) => Ok(()),
>             Err(error_code) => Err(error_code),
>         }
>     }
> }
> ```

We can ignore the `do_init` case(this is widely used by many MySQL clients or Drivers when they are connecting, here use empty is by design), only handle the `do_query` case is ok.

> if we change it to
> 
> ```diff
> async fn do_init(&mut self, database_name: &str) -> Result<()> {
> -       if database_name.is_empty() {
> -          return Ok(());
> -       }
>         let init_query = format!("USE `{}`;", database_name);
> 
> 
>         let do_query = self.do_query(&init_query).await;
>         match do_query {
>             Ok(_) => Ok(()),
>             Err(error_code) => Err(error_code),
>         }
>     }
> }
> ```
> 
> and check database name in `do_query`, **new connection will fail because new connection can be made without database name**(can see the test case https://github.com/datafuselabs/databend/blob/main/tests/suites/0_stateless/10_drivers/10_0000_python_mysql_driver.py)

Reply:
> > Actually, when database name is empty, it will return directy in `do_init` and `do_query` wont be called.
> > ```
> >   async fn do_init(&mut self, database_name: &str) -> Result<()> {
> >         if database_name.is_empty() {
> >             return Ok(());
> >         }
> >         let init_query = format!("USE `{}`;", database_name);
> > 
> > 
> >         let do_query = self.do_query(&init_query).await;
> >         match do_query {
> >             Ok(_) => Ok(()),
> >             Err(error_code) => Err(error_code),
> >         }
> >     }
> > }
> > ```
> 
> We can ignore the `do_init` case(this is widely used by many MySQL clients or Drivers when they are connecting, here use empty is by design), only handle the `do_query` case is ok.
> 
> > if we change it to
> > ```diff
> > async fn do_init(&mut self, database_name: &str) -> Result<()> {
> > -       if database_name.is_empty() {
> > -          return Ok(());
> > -       }
> >         let init_query = format!("USE `{}`;", database_name);
> > 
> > 
> >         let do_query = self.do_query(&init_query).await;
> >         match do_query {
> >             Ok(_) => Ok(()),
> >             Err(error_code) => Err(error_code),
> >         }
> >     }
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > and check database name in `do_query`, **new connection will fail because new connection can be made without database name**(can see the test case https://github.com/datafuselabs/databend/blob/main/tests/suites/0_stateless/10_drivers/10_0000_python_mysql_driver.py)

let me confirm the change to be made:

1. change `do_query` in `query/src/servers/mysql/mysql_interactive_worker.rs`: add check for empty database name;
2. leave `do_init` unchanged

but this cannot fix the [#4723](https://github.com/datafuselabs/databend/issues/4723) because when `USE ''` is received, `on_init` will be called and then `do_init`, and `OK(())` is return from `do_init`. And `do_query` wont be called.

Reply:
> Bye the way, I find other err:
> 
> In MySQL :
> 
> ```
> root@mysqldb 12:51:32 [(none)]> use ` `;
> ERROR 1102 (42000): Incorrect database name ' '
> root@mysqldb 12:51:34 [(none)]> select version();
> +-----------+
> | version() |
> +-----------+
> | 8.0.27    |
> +-----------+
> 1 row in set (0.00 sec)
> 
> root@mysqldb 12:51:41 [(none)]> use ` `;
> ERROR 1102 (42000): Incorrect database name ' '
> ```
> 
> In ClickHouse
> 
> ```
> ck :) select version();
> 
> SELECT version()
> 
> Query id: c5c52bf2-dc66-4bfa-a130-f02589627f0f
> 
> ┌─version()─┐
> │ 22.3.2.1  │
> └───────────┘
> 
> 1 rows in set. Elapsed: 0.001 sec. 
> 
> ck :) use ` `;
> 
> USE ` `
> 
> Query id: 8d08efed-0e6a-414c-8fe5-ff0f463ce39b
> 
> 
> 0 rows in set. Elapsed: 0.002 sec. 
> 
> Received exception from server (version 22.3.2):
> Code: 81. DB::Exception: Received from localhost:9876. DB::Exception: Database ` ` doesn't exist. (UNKNOWN_DATABASE)
> ```
> 
> But in databend's ClickHouse : the use query will hang. Could you create a new issue and fix it in a new pr?
> 
> ```
> databend :) select version();
> 
> SELECT version()
> 
> Query id: 64a37b8f-560d-4128-ad01-af8e67d00fc7
> 
> Connecting to localhost:9000 as user default.
> Connected to databend server version 2021.5.0 revision 54405.
> 
> ClickHouse client version is older than ClickHouse server. It may lack support for new features.
> 
> ┌─version()───────────────────────────────────────────────────────────────────────────────────┐
> │ DatabendQuery v0.2.1-alpha-6c170b4-simd(rust-1.61.0-nightly-2022-04-21T10:58:42.707348427Z) │
> └─────────────────────────────────────────────────────────────────────────────────────────────┘
> 
> 1 rows in set. Elapsed: 0.045 sec. 
> 
> databend :) use ` `;
> 
> USE ` `
> 
> Query id: 3229b05c-f707-4bfa-bad1-6de4985d7b5d
> ```
> 
> In a word, if use clickhouse-client connect databend server and try to use a not exists database, the connection will hang.

new issue is submitted [#5505](https://github.com/datafuselabs/databend/issues/5005)

Reply:
> Look this example:
> 
> ```
> >>> import mysql.connector
> >>> mydb = mysql.connector.connect(
> ...   host="127.0.0.1",
> ...   user="root",
> ...   passwd="",
> ...   port="3306")
> >>> 
> >>> mycursor = mydb.cursor()
> >>>  
> >>> mycursor.execute("SHOW TABLES")
> Traceback (most recent call last):
>   File "/home/taichong/.local/lib/python3.8/site-packages/mysql/connector/connection_cext.py", line 523, in cmd_query
>     self._cmysql.query(query,
> _mysql_connector.MySQLInterfaceError: No database selected
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/home/taichong/.local/lib/python3.8/site-packages/mysql/connector/cursor_cext.py", line 269, in execute
>     result = self._cnx.cmd_query(stmt, raw=self._raw,
>   File "/home/taichong/.local/lib/python3.8/site-packages/mysql/connector/connection_cext.py", line 528, in cmd_query
>     raise errors.get_mysql_exception(exc.errno, msg=exc.msg,
> mysql.connector.errors.ProgrammingError: 1046 (3D000): No database selected
> >>>  
> ```
> 
> If not set `db='xx'` the python driver will report a err when connect mysql.
> 
> So we could modify this test.

I am not sure whether this test case should be changed or not.

Do you think it's ok to call `on_query` rather than `on_init` when an `USE` command is received in `opensrv/mysql/src/lib.rs:run` ? Then we can check whether `USE` command is with empty database name or not. And `on_init` method should be changed as below:

```diff
-/// Called when client switches database.
+/// Called when a new client connection is accepted.
async fn on_init<'a>(
    &'a mut self,
    _: &'a str,
    _: InitWriter<'a, W>,
) -> Result<(), Self::Error> {
    Ok(())
}
```  


With this change, interactive `USE` command will call `on_query` and then `do_query` instead.